src/daemon.py

```diff
@@ -148,107 +170,77 @@ def update_metadata(job_hash: str, file_path: str, matches: List[str]) -> None:
     for v in current_meta.values():
         flat_meta.update(v)
 
-    logging.info("Fetched metadata: " + file_path)
-
-    pipe = redis.pipeline()
-    pipe.rpush(
+    redis.rpush(
         "meta:{}".format(job_hash),
         json.dumps({"file": file_path, "meta": flat_meta, "matches": matches}),
     )
-    pipe.hget("job:{}".format(job_hash), "total_files")
-    pipe.hincrby("job:{}".format(job_hash), "files_processed")
-    _, total_files, files_processed = pipe.execute()
-
-    if int(files_processed) >= int(total_files):
-        redis.hset("job:{}".format(job_hash), "status", "done")
 
 
-def execute_yara(job_hash: str, file: str) -> None:
-    if redis.hget("job:" + job_hash, "status") in [
+def execute_yara(job_hash: str, files: List[str]) -> None:
+    job_id = f"job:{job_hash}"
+    if redis.hget(job_id, "status") in [
         "cancelled",
         "failed",
     ]:
         return
 
-    rule = compile_yara(job_hash)
+    if len(files) == 0:
+        return
 
-    try:
-        matches = rule.match(data=open(file, "rb").read())
-    except yara.Error:
-        logging.exception("Yara failed to check file {}".format(file))
-        matches = None
-    except FileNotFoundError:
-        logging.exception(
-            "Failed to open file for yara check: {}".format(file)
-        )
-        matches = None
-
-    if matches:
-        logging.info("Processed (match): {}".format(file))
-        update_metadata(job_hash, file, [r.rule for r in matches])
-    else:
-        logging.info("Processed (nope ): {}".format(file))
+    rule = compile_yara(job_hash)
 
-        pipe = redis.pipeline()
-        pipe.rpush("false_positives:" + job_hash, file)
-        pipe.hget("job:{}".format(job_hash), "total_files")
-        pipe.hincrby("job:{}".format(job_hash), "files_processed")
-        _, total_files, files_processed = pipe.execute()
+    for sample in files:
+        try:
+            matches = rule.match(data=open(sample, "rb").read())
```

Comment:
I guess the FD gets closed after the yara matching right?

btw: the `match` function accepts filepaths as well so no need to read the data ourselves right?

Reply:
ref: https://yara.readthedocs.io/en/v3.4.0/yarapython.html#yara.Rules

Reply:
Thanks! I didn't notice this, and I think this is unnecessarily slowing down scanning.